src/slidedown/Slidedown.ts

```diff
@@ -0,0 +1,225 @@
+      this.categoryOptions.savingButtonText = SERVER_CONFIG_DEFAULTS_SLIDEDOWN.savingText;
+      this.categoryOptions.errorButtonText = this.categoryOptions.positiveUpdateButton;
+    }
+  }
+
+  async create(isInUpdateMode?: boolean) {
+    // TODO: dynamically change btns depending on if its first or repeat display of slidedown (subscribe vs update)
+    if (this.notificationIcons === null) {
+      const icons = await MainHelper.getNotificationIcons();
+
+      this.notificationIcons = icons;
+
+      // Remove any existing container
+      if (this.container.className.includes(SlidedownCssClasses.container)) {
+          removeDomElement(`#${SlidedownCssIds.container}`);
+      }
+      const positiveButtonText = isInUpdateMode && !!this.categoryOptions ?
+        this.categoryOptions.positiveUpdateButton : this.options.acceptButtonText;
+      const negativeButtonText = isInUpdateMode && !!this.categoryOptions ?
+        this.categoryOptions.negativeUpdateButton : this.options.cancelButtonText;
+      const messageText = isInUpdateMode && !!this.categoryOptions ?
+        this.categoryOptions.updateMessage : this.options.actionMessage;
+
+      const icon = this.getPlatformNotificationIcon();
+      const slidedownHtml = getSlidedownHtml({
+        messageText,
+        icon,
+        positiveButtonText,
+        negativeButtonText
+      });
+
+      // Insert the container
+      addDomElement('body', 'beforeend', `<div id="${SlidedownCssIds.container}"` +
+        `class="${SlidedownCssClasses.container} ${SlidedownCssClasses.reset}"></div>`);
+      // Insert the dialog
+      addDomElement(this.container, 'beforeend',
+          `<div id="${SlidedownCssIds.dialog}" class="${SlidedownCssClasses.dialog}">${slidedownHtml}</div>`);
+
+      // Animate it in depending on environment
+      addCssClass(this.container, bowser.mobile ? 'slide-up' : 'slide-down');
+
+      // Add click event handlers
+      this.allowButton.addEventListener('click', this.onSlidedownAllowed.bind(this));
+      this.cancelButton.addEventListener('click', this.onSlidedownCanceled.bind(this));
+      Event.trigger(Slidedown.EVENTS.SHOWN);
+    }
+  }
+
+  async onSlidedownAllowed(_: any) {
+    await Event.trigger(Slidedown.EVENTS.ALLOW_CLICK);
+  }
+
+  onSlidedownCanceled(_: any) {
+    Event.trigger(Slidedown.EVENTS.CANCEL_CLICK);
+    this.close();
+  }
+
+  close() {
+    addCssClass(this.container, 'close-slidedown');
+    once(this.dialog, 'animationend', (event: any, destroyListenerFn: () => void) => {
+      if (event.target === this.dialog &&
+          (event.animationName === 'slideDownExit' || event.animationName === 'slideUpExit')) {
+          // Uninstall the event listener for animationend
+          removeDomElement(`#${SlidedownCssIds.container}`);
+          destroyListenerFn();
+          Event.trigger(Slidedown.EVENTS.CLOSED);
+      }
+    }, true);
+  }
+
+  /**
+   * only used with Category Slidedown
+   */
+  toggleSaveState() {
```

Comment:
`toggleSaveState` seems it would be more fragile than just setting the state we want. It would be more declarative and won't have any possible issue of state carry over.

Reply:
Done.